### PR TITLE
Stop deployment from non-approved IP addresses

### LINKED ISSUE #1890: ACLs cannot be set when deploying storage accounts
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** develop @ 9e5b72a

## :no_entry_sign: Describe the problem

When deploying a new SRE, the desired ACLs cannot be set on the sensitive data storage account. This could be due to attempting to deploy from an IP address that is not listed under `admin_ip_addresses` in the SREConfig (only requests from these IP addresses are allowed to make changes to these storage accounts).

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
2024-05-14 16:24:18 [    INFO]  +  pulumi-python:dynamic:Resource sre_data_blob_egress_acl creating (4s) error: Exception calling application: Failed to set ACLs on storage account              _cmd.py:226
shgresrefersensitivedata.
2024-05-14 16:24:18 [    INFO]  +  pulumi-python:dynamic:Resource sre_data_blob_egress_acl **creating failed** error: Exception calling application: Failed to set ACLs on storage account              _cmd.py:226
shgresrefersensitivedata.
2024-05-14 16:24:18 [    INFO]  +  pulumi-python:dynamic:Resource sre_data_blob_ingress_acl creating (3s) error: Exception calling application: Failed to set ACLs on storage account             _cmd.py:226
shgresrefersensitivedata.
2024-05-14 16:24:18 [    INFO]  +  pulumi-python:dynamic:Resource sre_data_blob_ingress_acl **creating failed** error: Exception calling application: Failed to set ACLs on storage account             _cmd.py:226
shgresrefersensitivedata.
```

</details>

## :recycle: To reproduce

Deploy an SRE and observe this failure in the logs.

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :arrow_heading_up: Summary

ACLs can only be set/updated from approved admin IP addresses. This means that deploying from a non-admin IP address caused the deployment to fail. This PR catches the problem before deployment starts.

### :closed_umbrella: Related issues

Closes #1890

### :microscope: Tests

Tested locally.

### CODE REVIEW

File: data_safe_haven/functions/network.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import requests
+
+
+def current_ip_address(*, as_cidr: bool = False) -> str:
+    ip_address = requests.get("https://api.ipify.org", timeout=300).content.decode(
+        "utf8"
+    )
+    if as_cidr:
+        ip_address += "/32"
+    return ip_address
```
Comment (JimMadge) [inline data_safe_haven/functions/network.py]: Let's add some tests for this.

File: data_safe_haven/commands/sre.py [hunk 2]
```diff
@@ -55,6 +56,17 @@ def deploy(
...
+        ip_address = current_ip_address(as_cidr=True)
+        if ip_address not in sre_config.sre.admin_ip_addresses:
+            msg = " ".join(
+                [
+                    f"Current IP address {ip_address} is not authorised to deploy SRE '[green]{sre_config.description}[/]'.",
+                    "Please update the 'admin_ip_addresses' section of the config file.",
+                ]
+            )
+            raise DataSafeHavenConfigError(msg)
+
```
Comment (JimMadge) [inline data_safe_haven/commands/sre.py]: As there are two instances of this, can we put it in a function?

Do we want to raise `typer.Exit` here?

**Review by JimMadge — CHANGES_REQUESTED** [review]
Looks good.

Would want to improve the coverage and see if we can reduce duplicate code.

File: data_safe_haven/functions/network.py (hunk 1 above)
Comment (jemrobinson) [reply data_safe_haven/functions/network.py]: Done in 80fdf3a

File: data_safe_haven/commands/sre.py (hunk 2 above)
Comment (jemrobinson) [reply data_safe_haven/commands/sre.py]: Done in ac1ec310b.

I think it's unnecessary to raise a `typer.Exit`: this happens in a `try/except` block which will catch the error and raise `typer.Exit(1) from exc` after printing a "Could not deploy/teardown" message.

File: pyproject.toml [hunk 3]
```diff
@@ -116,6 +116,7 @@ dependencies = [
   "freezegun>=1.5",
   "pytest>=8.1",
   "pytest-mock>=3.14",
+  "requests-mock>=1.12",
```
Comment (JimMadge) [inline pyproject.toml:119]: Good find!

**Review by JimMadge — APPROVED** [review]

**Pull request merged**